src/server/pps/server/poller.go

```diff
@@ -119,27 +147,133 @@ func (a *apiServer) pollPipelines(pachClient *client.APIClient) {
+		m.eventCh <- &pipelineEvent{eventType: writeEv, pipeline: pipeline}
 
-		// move to next pipeline
+		// move to next pipeline (after 2s sleep)
 		return backoff.ErrContinue
-	}, backoff.NewConstantBackOff(2*time.Second),
+	}, backoff.NewConstantBackOff(pollBackoffTime),
 		backoff.NotifyContinue("pollPipelines"),
 	); err != nil {
 		if ctx.Err() == nil {
 			panic("pollPipelines is exiting prematurely which should not happen; restarting pod...")
 		}
 	}
 }
+
+func (m *ppsMaster) pollPipelinePods(pollClient *client.APIClient) {
+	ctx := pollClient.Ctx()
+	if err := backoff.RetryUntilCancel(ctx, func() error {
+		// watchChan will be nil if the Watch call below errors, this means
+		// that we won't receive events from k8s and won't be able to detect
+		// errors in pods. We could just return that error and retry but that
+		// prevents pachyderm from creating pipelines when there's an issue
+		// talking to k8s.
+		kubePipelineWatch, err := m.a.env.GetKubeClient().CoreV1().Pods(m.a.namespace).Watch(
+			metav1.ListOptions{
+				LabelSelector: metav1.FormatLabelSelector(metav1.SetAsLabelSelector(
+					map[string]string{
+						"component": "worker",
+					})),
+				Watch: true,
+			})
+		if err != nil {
+			return errors.Wrap(err, "failed to watch kubernetes pods")
+		}
+		defer kubePipelineWatch.Stop()
+		for event := range kubePipelineWatch.ResultChan() {
+			// if we get an error we restart the watch
+			if event.Type == kube_watch.Error {
+				return errors.Wrap(kube_err.FromObject(event.Object), "error while watching kubernetes pods")
+			} else if event.Type == "" {
+				// k8s watches seem to sometimes get stuck in a loop returning events
+				// with Type = "". We treat these as errors as otherwise we get an
+				// endless stream of them and can't do anything.
+				return errors.New("error while watching kubernetes pods: empty event type")
+			}
+			pod, ok := event.Object.(*v1.Pod)
+			if !ok {
+				continue // irrelevant event
+			}
+			if pod.Status.Phase == v1.PodFailed {
+				log.Errorf("pod failed because: %s", pod.Status.Message)
+			}
+			pipelineName := pod.ObjectMeta.Annotations["pipelineName"]
+			for _, status := range pod.Status.ContainerStatuses {
+				if status.State.Waiting != nil && failures[status.State.Waiting.Reason] {
+					if err := m.a.setPipelineCrashing(ctx, pipelineName, status.State.Waiting.Message); err != nil {
+						return errors.Wrap(err, "error moving pipeline to CRASHING")
+					}
+				}
+			}
+			for _, condition := range pod.Status.Conditions {
+				if condition.Type == v1.PodScheduled &&
+					condition.Status != v1.ConditionTrue && failures[condition.Reason] {
+					if err := m.a.setPipelineCrashing(ctx, pipelineName, condition.Message); err != nil {
+						return errors.Wrap(err, "error moving pipeline to CRASHING")
+					}
+				}
+			}
+		}
+		return backoff.ErrContinue // keep polling until cancelled (RetryUntilCancel)
+	}, &backoff.ZeroBackOff{}, backoff.NotifyContinue("pollPipelinePods"),
```

Comment:
similar to above, is there a reason to make this instant rather than the exponential infinite backoff it was before? not sure the failure modes here

Reply:
I'm not totally sure what the right answer is either; I thought about this some more too and changed it back to exponential backoff because this code is making an RPC, so if the `Watch` RPC is failing because k8s API server is overloaded, then at least pachd won't be contributing to the overload.